[AArch64] Combine store (trunc X to <3 x i8>) to sequence of ST1.b.

fhahn opened the pull request:
Improve codegen for (trunc X to <3 x i8>) by converting it to a sequence of 3 ST1.b, but first converting the truncate operand to either v8i8 or v16i8, extracting the lanes for the truncate results and storing them.

At the moment, there are almost no cases in which such vector operations will be generated automatically. The motivating case is non-power-of-2 SLP vectorization: https://github.com/llvm/llvm-project/pull/77790

fhahn committed:
[AArch64] Combine store (trunc X to <3 x i8>) to sequence of ST1.b.

Improve codegen for (trunc X to <3 x i8>) by converting it to a sequence
of 3 ST1.b, but first converting the truncate operand to either v8i8 or
v16i8, extracting the lanes for the truncate results and storing them.

At the moment, there are almost no cases in which such vector operations
will be generated automatically. The motivating case is non-power-of-2
SLP vectorization: https://github.com/llvm/llvm-project/pull/77790

efriedma-quic commented on llvm/test/CodeGen/AArch64/vec3-loads-ext-trunc-stores.ll:
```diff
@@ -154,17 +154,12 @@ define <3 x i32> @load_v3i32(ptr %src) {
-; CHECK-NEXT:    add sp, sp, #16
+; CHECK-NEXT:    add x8, x0, #4
+; CHECK-NEXT:    ld1r.4h { v0 }, [x8]
+; CHECK-NEXT:    ldr w8, [x0]
```
This isn't really doing a good job of demonstrating what you're trying to do here... maybe add a testcase with some arithmetic, so the store and the load can't be combined together?

Also, this looks like it's getting miscompiled; it's only storing two bytes.

efriedma-quic replied:
> The store and load have different addresses, so it shouldn't be possible to combine them

I mean, the load involves some INSERT_VECTOR_ELT, and the store involves some EXTRACT_VECTOR_ELT, and the two are getting combined together, so it's not a "pure" store sequence; one of the stores is actually from an integer register.  If the combiners get smarter, you might end up not using vector ops at all.

fhahn replied:
Ah I see, thanks! Added additional variants in https://github.com/llvm/llvm-project/commit/e7b4ff8119403509da3e7941dcb86b1c6a6d61c5 and updated the PR

fhahn committed:
!fixup fix extract index computation, adjust pointer info and alignment by offset

fhahn committed:
Merge branch 'main' into aarch64-lower-store-trunc-v3i8

fhahn committed:
!fixup add assert that offset is undef.

fhahn committed:
Merge branch 'main' into aarch64-lower-store-trunc-v3i8

fhahn committed:
!fixup update additional tests.

fhahn committed:
Merge branch 'main' into aarch64-lower-store-trunc-v3i8

fhahn committed:
!fixup update on top of new test coverage.

Update checks after adding more tests in e7b4ff8119403509da3e7941dcb86b1c6a6d61c5

efriedma-quic commented on llvm/test/CodeGen/AArch64/vec3-loads-ext-trunc-stores.ll:
```diff
@@ -281,23 +279,19 @@ entry:
-; CHECK-NEXT:    xtn.8b v1, v0
-; CHECK-NEXT:    umov.h w8, v0[2]
-; CHECK-NEXT:    str s1, [sp, #12]
-; CHECK-NEXT:    ldrh w9, [sp, #12]
```
Spent a little time looking at why the default code is so horrible; the primary issue is actually the way the legalizer (GenWidenVectorStores) is trying to lower the operation into an i16 store followed by an i8 store. It ends up generating a bitcast from v4i8 to v2i16, and the default handling for that is completely terrible (it doesn't know how to use a shuffle, so it goes through a stack temporary).

Maybe worth looking into improving the bitcast situation if you're going to continue looking at very narrow vector types.

fhahn replied:
Yep, I am planning to look into this as well.

fhahn replied:
Actually, I am not entirely sure what the exact issue is; the bit cast itself should be a no-op. Perhaps the issue is that the vector types aren't legal? 

It seems like we are going from v3i8 -> v4i8, which isn't legal, and then going from v4i8 -> v4i16 which is legal. Perhaps it would be better to go from v3i8 to v8i8, but I couldn't find a way to do that; setting custom actions require from & to types to by MVTs, but v3i8 isn't a simple value type unfortunately.

efriedma-quic replied:
v4i8 and v2i16 aren't legal types.  And the legalization rule we use for them is "promote", so the bitcast isn't a no-op. There's padding between the elements.  So we need to do a shuffle... but neither AArch64TargetLowering::ReplaceBITCASTResults nor DAGTypeLegalizer::PromoteIntRes_BITCAST has code to do that, so we end up using CreateStackStoreLoad.

The legalization rule for vector types is under the control of the target; see AArch64TargetLowering::getPreferredVectorAction.  You can use that to force v4i8->v8i8.  But changing that impacts a lot of code.

fhahn replied:
Thanks, I'll look into that separately.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21471,6 +21471,57 @@ bool isHalvingTruncateOfLegalScalableType(EVT SrcVT, EVT DstVT) {
+      ISD::BITCAST, DL, WideVT.getSizeInBits() == 64 ? MVT::v8i8 : MVT::v16i8,
+      WideTrunc);
+
+  unsigned IdxScale = WideVT.getScalarSizeInBits() / 8;
```
Instead of writing this out, can we just use TargetLowering::scalarizeVectorStore?  I think it does roughly the same thing.

fhahn replied:
I am not sure, I tried but the code here combines the trunc with the stores; it looks like`scalarizeVectorStore` would generate a 16 bit store and a 8 bit store, and we first need to do the cast separately.

efriedma-quic replied:
If it doesn't work, that's fine

fhahn commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21471,6 +21471,57 @@ bool isHalvingTruncateOfLegalScalableType(EVT SrcVT, EVT DstVT) {
+  EVT ValueVT = Value.getValueType();
+
+  if (ST->isVolatile() || !Subtarget->isLittleEndian() ||
+      ST->getOriginalAlign() >= 4 || Value.getOpcode() != ISD::TRUNCATE ||
```
Yes, this was only relevant for the load case I think. I will remove it (and also add tests with different alignments)

fhahn replied:
Removed the check and added now tests with different alignments; the alignment impacts load codegen only.

fhahn committed:
Merge branch 'main' into aarch64-lower-store-trunc-v3i8

fhahn committed:
!fixup remove alignment check, update after new tests, use MMO.

fhahn commented:
Updated, comments should be addressed and also updated to use `MachineFunction::getMachineMemOperand`

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21471,6 +21471,53 @@ bool isHalvingTruncateOfLegalScalableType(EVT SrcVT, EVT DstVT) {
+
+  SDValue E0 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, DL, MVT::i8, Cast,
+                           DAG.getConstant(0, DL, MVT::i64));
+  Chain = DAG.getStore(Chain, DL, E0, ST->getBasePtr(), ST->getMemOperand());
```
Missing getMachineMemOperand call here?  (The MachineMemOperand for the original store has the wrong size.)

fhahn replied:
Updated, thanks!

fhahn committed:
!fixup also adjust mem-operand of store to base address.

efriedma-quic reviewed (approved):
LGTM
